electrumx/server/session.py

```diff
@@ -1186,6 +1188,39 @@ async def block_headers(self, start_height, count, cp_height=0):
         self.bump_cost(cost)
         return result
 
+    async def block_headers_array(self, start_height, count, cp_height=0):
+        '''Return block headers in an array for the main chain;
+        starting at start_height.
+
+        start_height and count must be non-negative integers.  At most
+        MAX_CHUNK_SIZE headers will be returned.
+        '''
+        start_height = non_negative_integer(start_height)
+        count = non_negative_integer(count)
+        cp_height = non_negative_integer(cp_height)
+        cost = count / 50
+
+        max_size = self.MAX_CHUNK_SIZE
+        count = min(count, max_size)
+        headers, count = await self.db.read_headers(start_height, count)
+        cursor = 0
+        result = {'count': count, 'max': max_size, 'headers': []}
+        if count and cp_height:
+            cost += 1.0
+            last_height = start_height + count - 1
+            result.update(await self._merkle_proof(cp_height, last_height))
+
+        height = 0
+        while cursor < len(headers):
+            header_size = self.db.dynamic_header_len(height)
+            header = headers[cursor:cursor + header_size]
+            result['headers'].append(header)
+            cursor += header_size
+            height += 1
```

Comment:
`dynamic_header_len` will yield an `assert` error for chains with static header length.  It's also unnecessarily expensive to use it for this, because its implementation checks both the start and end positions of each header, so this code `cursor + header_size` is actually doing the redundant calculation `cursor + self.dynamic_header_offset(height + 1) - cursor`.

I'd suggest replacing `header_size` with `next_cursor`, setting it to `self.db.header_offset` (which works for both static and dynamic lengths), and changing the slicing/incrementing accordingly.